module/zfs/arc.c

```diff
@@ -5611,6 +5611,53 @@ arc_access(arc_buf_hdr_t *hdr, kmutex_t *hash_lock)
 	}
 }
 
+/*
+ * This routine is called by dbuf_hold() to update the arc_access()
+ * information which otherwise would be skipped for dbufs in the cache.
+ */
+void
+arc_buf_access(arc_buf_t *buf)
+{
+	arc_buf_hdr_t *hdr;
+	kmutex_t *hash_lock;
+
+	mutex_enter(&buf->b_evict_lock);
+	hdr = buf->b_hdr;
+
+	if (hdr->b_l1hdr.b_state == arc_anon || HDR_EMPTY(hdr)) {
+		mutex_exit(&buf->b_evict_lock);
+		return;
+	}
+
+	/*
+	 * Because the hash_lock is obtained from the arc_buf_hdr_t it is
+	 * possible the arc_buf_hdr_t may be released before the hash_lock
+	 * can be acquired.  Detect this case by verifying the arc_buf_hdr_t
+	 * is still hashed after taking the lock.
+	 */
+	hash_lock = HDR_LOCK(buf->b_hdr);
+	mutex_enter(hash_lock);
+
+	if (!HDR_IN_HASH_TABLE(hdr)) {
```

Comment:
Did you actually hit this case while testing? I'm trying to understand how a `dbuf` with data associated with it would not be in the hash table. I don't think we can evict it prior to someone calling `remove_reference`

Reply:
Yes, I did.  It comes up pretty quickly when running the ZTS.  What I think can happen here is we start with a dbuf which has data in it but before we can take the hash lock it gets released.  This appears to be possible because I'm using `HDR_LOCK` to get the hash lock and `buf_discard_identity()` is called outside the `b_evict_lock` in `arc_released()`.  I'd love a better solution for this if you have one.